## **Initiation**

**Importing the required libraries**

In [34]:
import pandas as pd
import numpy as np

**Reading in the dataframe (sourced from X(Twitter) Data Challenge Space)**

In [35]:
pd.set_option('display.max_columns', None)      # to display all the columns present in the dataframe
data = pd.read_csv('/content/fifa21 data.csv')

<ipython-input-35-f9f0a2650cdf>:2: DtypeWarning: Columns (76) have mixed types. Specify dtype option on import or set low_memory=False.
  data = pd.read_csv('/content/fifa21 data.csv')


In [36]:
data.sample(3)

ID          Name           LongName  \
14736  251632     N. Tarnat      Niklas Tarnat   
13046  237337  T. Verlinden  Thibaud Verlinden   
11928  167925    J. O'Brien       Joey O'Brien   

                                               photoUrl  \
14736  https://cdn.sofifa.com/players/251/632/21_60.png   
13046  https://cdn.sofifa.com/players/237/337/21_60.png   
11928  https://cdn.sofifa.com/players/167/925/21_60.png   

                                               playerUrl          Nationality  \
14736  http://sofifa.com/player/251632/niklas-tarnat/...              Germany   
13046  http://sofifa.com/player/237337/thibaud-verlin...              Belgium   
11928  http://sofifa.com/player/167925/joey-obrien/21...  Republic of Ireland   

       Age  ↓OVA  POT                     Club     Contract   Positions  \
14736   22    61   72      \n\n\n\nHannover 96  2019 ~ 2021         CDM   
13046   20    62   75       \n\n\n\nStoke City  2016 ~ 2022      RW, ST   
11928   34    64   64  \n\n\n\nShamrock Rovers  2018 ~ 2025  CB, RB, LB   

      Height Weight Preferred Foot  BOV Best Position     Joined  \
14736  179cm   73kg          Right   62            CM  01-Jul-19   
13046  173cm   65kg          Right   64            RM  14-Jul-16   
11928  188cm   82kg          Right   64            CB  27-Jan-18   

      Loan Date End  Value  Wage Release Clause  Attacking  Crossing  \
14736           NaN  €700K   €4K          €720K        270        59   
13046           NaN  €950K   €4K          €1.5M        273        64   
11928           NaN  €190K  €900          €264K        274        60   

       Finishing  Heading Accuracy  Short Passing  Volleys  Skill  Dribbling  \
14736         53                53             59       46    292         56   
13046         50                46             60       53    286         69   
11928         41                61             63       49    281         57   

       Curve  FK Accuracy  Long Passing  Ball Control  Movement  Acceleration  \
14736     50           58            65            63       328            64   
13046     55           42            55            65       352            71   
11928     63           37            62            62       222            32   

       Sprint Speed  Agility  Reactions  Balance  Power  Shot Power  Jumping  \
14736            64       67         61       72    314          66       69   
13046            74       75         51       81    254          69       46   
11928            36       38         64       52    277          47       66   

       Stamina  Strength  Long Shots  Mentality  Aggression  Interceptions  \
14736       66        54          59        288          64             57   
13046       54        36          49        225          39             20   
11928       55        70          39        288          64             64   

       Positioning  Vision  Penalties  Composure  Defending  Marking  \
14736           53      55         59         52        173       57   
13046           57      59         50         59        134       48   
11928           52      53         55         65        191       65   

       Standing Tackle  Sliding Tackle  Goalkeeping  GK Diving  GK Handling  \
14736               57              59           52         11           12   
13046               37              49           53          9            7   
11928               64              62           48         15            7   

       GK Kicking  GK Positioning  GK Reflexes  Total Stats  Base Stats  W/F  \
14736          13               7            9         1717         357  3 ★   
13046          10              15           12         1577         335  3 ★   
11928           8               8           10         1581         322  4 ★   

       SM     A/W     D/W   IR  PAC  SHO  PAS  DRI  DEF  PHY Hits  
14736  2★  Medium  Medium  1 ★   64   57   59   60   57   60    6  
13046  3★  Medium  Medium  1 ★   73   54  

In [37]:
data.shape    #displays a tuple (number of rows, number of columns)

(18979, 77)

In [38]:
data.info()     #detailed information about the columns of the dataframe

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 18979 entries, 0 to 18978
Data columns (total 77 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   ID                18979 non-null  int64 
 1   Name              18979 non-null  object
 2   LongName          18979 non-null  object
 3   photoUrl          18979 non-null  object
 4   playerUrl         18979 non-null  object
 5   Nationality       18979 non-null  object
 6   Age               18979 non-null  int64 
 7   ↓OVA              18979 non-null  int64 
 8   POT               18979 non-null  int64 
 9   Club              18979 non-null  object
 10  Contract          18979 non-null  object
 11  Positions         18979 non-null  object
 12  Height            18979 non-null  object
 13  Weight            18979 non-null  object
 14  Preferred Foot    18979 non-null  object
 15  BOV               18979 non-null  int64 
 16  Best Position     18979 non-null  object
 17  Joined      

**Create of copy of DataFrame**

In [39]:
fifa = data.copy()    #so that original dataset remains unaltered

# **Data cleaning**

In [40]:
fifa['Club'].unique()

array(['\n\n\n\nFC Barcelona', '\n\n\n\nJuventus',
       '\n\n\n\nAtlético Madrid', '\n\n\n\nManchester City',
       '\n\n\n\nParis Saint-Germain', '\n\n\n\nFC Bayern München',
       '\n\n\n\nLiverpool', '\n\n\n\nReal Madrid', '\n\n\n\nChelsea',
       '\n\n\n\nTottenham Hotspur', '\n\n\n\nInter', '\n\n\n\nNapoli',
       '\n\n\n\nBorussia Dortmund', '\n\n\n\nManchester United',
       '\n\n\n\nArsenal', '\n\n\n\nLazio', '\n\n\n\nLeicester City',
       '\n\n\n\nBorussia Mönchengladbach', '\n\n\n\nReal Sociedad',
       '\n\n\n\nAtalanta', '\n\n\n\nOlympique Lyonnais', '\n\n\n\nMilan',
       '\n\n\n\nVillarreal CF', '\n\n\n\nRB Leipzig', '\n\n\n\nCagliari',
       '\n\n\n\nAjax', '\n\n\n\nSL Benfica', '\n\n\n\nAS Monaco',
       '\n\n\n\nWolverhampton Wanderers', '\n\n\n\nEverton',
       '\n\n\n\nFiorentina', '\n\n\n\nFC Porto', '\n\n\n\nRC Celta',
       '\n\n\n\nTorino', '\n\n\n\nSevilla FC', '\n\n\n\nGrêmio',
       '\n\n\n\nReal Betis', '\n\n\n\nRoma', '\n\n\n\nNewcastle Unite

**Getting rid of leading spaces in 'Club' column**

In [41]:
fifa['Club'] = fifa['Club'].str.strip()

In [42]:
fifa['Club'].unique()

array(['FC Barcelona', 'Juventus', 'Atlético Madrid', 'Manchester City',
       'Paris Saint-Germain', 'FC Bayern München', 'Liverpool',
       'Real Madrid', 'Chelsea', 'Tottenham Hotspur', 'Inter', 'Napoli',
       'Borussia Dortmund', 'Manchester United', 'Arsenal', 'Lazio',
       'Leicester City', 'Borussia Mönchengladbach', 'Real Sociedad',
       'Atalanta', 'Olympique Lyonnais', 'Milan', 'Villarreal CF',
       'RB Leipzig', 'Cagliari', 'Ajax', 'SL Benfica', 'AS Monaco',
       'Wolverhampton Wanderers', 'Everton', 'Fiorentina', 'FC Porto',
       'RC Celta', 'Torino', 'Sevilla FC', 'Grêmio', 'Real Betis', 'Roma',
       'Newcastle United', 'Eintracht Frankfurt', 'Valencia CF',
       'Medipol Başakşehir FK', 'Inter Miami', 'Bayer 04 Leverkusen',
       'Levante UD', 'Crystal Palace', 'Athletic Club de Bilbao',
       'Shanghai SIPG FC', 'VfL Wolfsburg',
       'Guangzhou Evergrande Taobao FC', 'Al Shabab',
       'Olympique de Marseille', 'Los Angeles FC',
       'Beijing Sino

In [43]:
fifa['Contract'].unique()

array(['2004 ~ 2021', '2018 ~ 2022', '2014 ~ 2023', '2015 ~ 2023',
       '2017 ~ 2022', '2017 ~ 2023', '2018 ~ 2024', '2014 ~ 2022',
       '2018 ~ 2023', '2016 ~ 2023', '2013 ~ 2023', '2011 ~ 2023',
       '2009 ~ 2022', '2005 ~ 2021', '2011 ~ 2021', '2015 ~ 2022',
       '2017 ~ 2024', '2010 ~ 2024', '2012 ~ 2021', '2019 ~ 2024',
       '2015 ~ 2024', '2017 ~ 2025', '2020 ~ 2025', '2019 ~ 2023',
       '2008 ~ 2023', '2015 ~ 2021', '2020 ~ 2022', '2012 ~ 2022',
       '2016 ~ 2025', '2013 ~ 2022', '2011 ~ 2022', '2012 ~ 2024',
       '2016 ~ 2021', '2012 ~ 2023', '2008 ~ 2022', '2019 ~ 2022',
       '2017 ~ 2021', '2013 ~ 2024', '2020 ~ 2024', '2010 ~ 2022',
       '2020 ~ 2021', '2011 ~ 2024', '2020 ~ 2023', '2014 ~ 2024',
       '2013 ~ 2026', '2016 ~ 2022', '2010 ~ 2021', '2013 ~ 2021',
       '2019 ~ 2025', '2018 ~ 2025', '2016 ~ 2024', '2018 ~ 2021',
       '2009 ~ 2024', '2007 ~ 2022', 'Jun 30, 2021 On Loan',
       '2009 ~ 2021', '2019 ~ 2021', '2019 ~ 2026', 'Free', '2012 ~ 

**Formatting the 'Contract' column by creating 3 new columns as - Contract start, contract end, and contract length(years)**

In [44]:
def extract_contract_info(contract):
	if contract == 'Free' or 'On Loan' in contract:
		start_date = np.nan
		end_date = np.nan
		contract_length = 0
	else:
		start_date, end_date = contract.split(' ~ ')
		start_year = int(start_date[:4])
		end_year = int(end_date[:4])
		contract_length = end_year - start_year
	return start_date, end_date, contract_length

**Applying the function created for formatting of 'Contract' column to the dataframe**

In [45]:
new_cols = ['Contract start', 'contract end', 'contract length(years)']
new_data = fifa['Contract'].apply(lambda x: pd.Series(extract_contract_info(x)))

**Inserting new columns to the dataframe 'Fifa'**

In [46]:
for i in range(len(new_cols)):
	fifa.insert(loc=fifa.columns.get_loc('Contract')+1+i, column = new_cols[i], value=new_data[i])

In [47]:
fifa.sample(3)

ID          Name        LongName  \
1649   152912       J. Sand       José Sand   
9125   239341  L. Hemmerich  Luke Hemmerich   
16933  247179     L. Müller     Leon Müller   

                                               photoUrl  \
1649   https://cdn.sofifa.com/players/152/912/21_60.png   
9125   https://cdn.sofifa.com/players/239/341/21_60.png   
16933  https://cdn.sofifa.com/players/247/179/21_60.png   

                                               playerUrl Nationality  Age  \
1649   http://sofifa.com/player/152912/jose-sand/210006/   Argentina   39   
9125   http://sofifa.com/player/239341/luke-hemmerich...     Germany   22   
16933  http://sofifa.com/player/247179/leon-muller/21...     Germany   19   

       ↓OVA  POT                   Club     Contract Contract start  \
1649     75   75    Club Atlético Lanús  2019 ~ 2021           2019   
9125     66   77  FC Würzburger Kickers  2019 ~ 2021           2019   
16933    57   72        SV Darmstadt 98  2019 ~ 2022           2019   

      contract end  contract length(years) Positions Height Weight  \
1649          2021                     2.0        ST  182cm   79kg   
9125          2021                     2.0    RB, RM  180cm   70kg   
16933         2022                     3.0   CAM, CM  180cm   75kg   

      Preferred Foot  BOV Best Position     Joined Loan Date End  Value  Wage  \
1649           Right   75            ST  21-Jan-19           NaN  €1.9M  €15K   
9125           Right   66            RB  01-Jul-19           NaN  €1.8M   €4K   
16933          Right   59           CAM  29-Jan-19           NaN  €400K   €2K   

      Release Clause  Attacking  Crossing  Finishing  Heading Accuracy  \
1649           €2.9M        357        53         81                77   
9125           €1.8M        273        63         49                55   
16933          €490K        240        48         55                35   

       Short Passing  Volleys  Skill  Dribbling  Curve  FK Accuracy  \
1649              69       77    311         67     63           63   
9125              62       44    313         63     70           57   
16933             60       42    282         58     49           56   

       Long Passing  Ball Control  Movement  Acceleration  Sprint Speed  \
1649             43            75       248            49            53   
9125             58            65       347            75            79   
16933            58            61       294            62            54   

       Agility  Reactions  Balance  Power  Shot Power  Jumping  Stamina  \
1649        38         81       27    353          77       71       49   
9125        65         61       67    352          78       70       85   
16933       65         44       69    256          61       41       34   

       Strength  Long Shots  Mentality  Aggression  Interceptions  \
1649         83          73        343          73             33   
9125         68          51        304          77             54   
16933        62          58        240          48             39   

       Positioning  Vision  Penalties  Composure  Defending  Marking  \
1649            83      69         85         85         91       45   
9125            58      59         56         55        173       51   
16933           51      56         46         44        127       34   

       Standing Tackle  Sliding Tackle  Goalkeeping  GK Diving  GK Handling  \
1649                21              25           65         11           16   
9125                60              62           51          9            9   
16933               49              44           47          9            8   

       GK Kicking  GK Positioning  GK Reflexes  Total Stats  Base Stats  W/F  \
1649           15               9           14         1768         365  4 ★   
9125           10               9           14         1813         388  2 ★   
16933          14              11            5         1486         321  3 ★  

**Insert a new column 'Contract Status' based on the status of the contract fetched from 'Contract' column**

In [48]:
def categorize_contract_status(contract):
	if contract == 'Free':
		return 'Free'
	elif 'On Loan' in contract:
		return 'On Loan'
	else:
		return 'Contract'

In [49]:
fifa.insert(fifa.columns.get_loc('Contract')+1, 'Contract Status', fifa['Contract'].apply(categorize_contract_status))

In [50]:
fifa.sample(3)

ID          Name         LongName  \
14572  246451  V. Triboulet  Vinni Triboulet   
12214  255302      G. Firpo  Guillermo Firpo   
16275  254243   E. Anderson  Elliot Anderson   

                                               photoUrl  \
14572  https://cdn.sofifa.com/players/246/451/21_60.png   
12214  https://cdn.sofifa.com/players/255/302/21_60.png   
16275  https://cdn.sofifa.com/players/254/243/21_60.png   

                                               playerUrl Nationality  Age  \
14572  http://sofifa.com/player/246451/vinni-triboule...      France   21   
12214  http://sofifa.com/player/255302/guillermo-firp...     Uruguay   31   
16275  http://sofifa.com/player/254243/elliot-anderso...    Scotland   17   

       ↓OVA  POT                   Club     Contract Contract Status  \
14572    61   72      AS Nancy Lorraine  2018 ~ 2022        Contract   
12214    63   63  Liverpool Fútbol Club  2020 ~ 2020        Contract   
16275    58   80       Newcastle United  2019 ~ 2021        Contract   

      Contract start contract end  contract length(years) Positions Height  \
14572           2018         2022                     4.0    ST, LW  183cm   
12214           2020         2020                     0.0       CDM  186cm   
16275           2019         2021                     2.0   CAM, LM  170cm   

      Weight Preferred Foot  BOV Best Position     Joined Loan Date End  \
14572   75kg          Right   63            ST  30-Nov-18           NaN   
12214   82kg          Right   63           CDM  21-Jan-20           NaN   
16275   60kg          Right   60           CAM  08-Nov-19           NaN   

       Value  Wage Release Clause  Attacking  Crossing  Finishing  \
14572  €750K   €1K          €878K        271        38         65   
12214  €325K  €500          €591K        201        37         33   
16275  €650K  €750          €798K        266        54         55   

       Heading Accuracy  Short Passing  Volleys  Skill  Dribbling  Curve  \
14572                57             57       54    228         61     46   
12214                43             59       29    242         48     33   
16275                40             62       55    284         64     57   

       FK Accuracy  Long Passing  Ball Control  Movement  Acceleration  \
14572           32            33            56       324            76   
12214           31            66            64       276            57   
16275           49            53            61       321            70   

       Sprint Speed  Agility  Reactions  Balance  Power  Shot Power  Jumping  \
14572            66       63         57       62    296          62       68   
12214            51       59         62       47    249          41       52   
16275            65       66         43       77    228          60       44   

       Stamina  Strength  Long Shots  Mentality  Aggression  Interceptions  \
14572       54        56          56        233          38             16   
12214       58        70          28        263          67             54   
16275       48        33          43        223          34             23   

       Positioning  Vision  Penalties  Composure  Defending  Marking  \
14572           64      54         61         57         52       21   
12214           38      62         42         44        188       55   
16275           50      58         58         54         94       33   

       Standing Tackle  Sliding Tackle  Goalkeeping  GK Diving  GK Handling  \
14572               16              15           47         14            5   
12214               69              64           51          6           14   
16275               31              30           40          6           12   

       GK Kicking  GK Positioning  GK Reflexes  Total Stats  Base Stats  W/F  \
14572          10               9            9         1451         315  3 ★   
12214           7              14           10         1470         322  3 ★   
16275     

In [51]:
fifa['Contract Status'].unique()

array(['Contract', 'On Loan', 'Free'], dtype=object)

In [52]:
fifa['Height'].unique()

array(['170cm', '187cm', '188cm', '181cm', '175cm', '184cm', '191cm',
       '178cm', '193cm', '185cm', '199cm', '173cm', '168cm', '176cm',
       '177cm', '183cm', '180cm', '189cm', '179cm', '195cm', '172cm',
       '182cm', '186cm', '192cm', '165cm', '194cm', '167cm', '196cm',
       '163cm', '190cm', '174cm', '169cm', '171cm', '197cm', '200cm',
       '166cm', '6\'2"', '164cm', '198cm', '6\'3"', '6\'5"', '5\'11"',
       '6\'4"', '6\'1"', '6\'0"', '5\'10"', '5\'9"', '5\'6"', '5\'7"',
       '5\'4"', '201cm', '158cm', '162cm', '161cm', '160cm', '203cm',
       '157cm', '156cm', '202cm', '159cm', '206cm', '155cm'], dtype=object)

**Converting the 'Height' column into uniform values in centimeters**

In [53]:
def convert_height(height):
	if 'cm' in height:
		return int(height.strip('cm'))
	else:
		feet, inches = height.split("'")
		total_inches = int(feet)*12 + int(inches.strip('"'))
		return round(total_inches*2.54)

In [54]:
fifa['Height'] = fifa['Height'].apply(convert_height)
fifa['Height'].unique()

array([170, 187, 188, 181, 175, 184, 191, 178, 193, 185, 199, 173, 168,
       176, 177, 183, 180, 189, 179, 195, 172, 182, 186, 192, 165, 194,
       167, 196, 163, 190, 174, 169, 171, 197, 200, 166, 164, 198, 201,
       158, 162, 161, 160, 203, 157, 156, 202, 159, 206, 155])

**Renaming the 'Height' column as 'Height(cms)'**

In [55]:
fifa = fifa.rename(columns={'Height':'Height(cms)'})

In [56]:
fifa.sample(3)

ID           Name          LongName  \
7979   257937    B. Süleyman    Burak Süleyman   
12982  212167      C. Vernam    Charles Vernam   
15600  234259  L. Hardcastle  Lewis Hardcastle   

                                               photoUrl  \
7979   https://cdn.sofifa.com/players/257/937/21_60.png   
12982  https://cdn.sofifa.com/players/212/167/21_60.png   
15600  https://cdn.sofifa.com/players/234/259/21_60.png   

                                               playerUrl Nationality  Age  \
7979   http://sofifa.com/player/257937/burak-suleyman...      Turkey   25   
12982  http://sofifa.com/player/212167/charles-vernam...     England   23   
15600  http://sofifa.com/player/234259/lewis-hardcast...     England   21   

       ↓OVA  POT           Club     Contract Contract Status Contract start  \
7979     67   68     Göztepe SK  2020 ~ 2022        Contract           2020   
12982    63   69  Burton Albion  2020 ~ 2022        Contract           2020   
15600    59   67         Barrow  2019 ~ 2021        Contract           2019   

      contract end  contract length(years) Positions  Height(cms) Weight  \
7979          2022                     2.0   LM, CAM          169   65kg   
12982         2022                     2.0        LW          170   74kg   
15600         2021                     2.0   CM, CAM          175   70kg   

      Preferred Foot  BOV Best Position     Joined Loan Date End  Value Wage  \
7979           Right   67            LM  08-Aug-20           NaN  €1.2M  €8K   
12982          Right   64            RM  20-Jul-20           NaN  €750K  €2K   
15600          Right   62           CAM  28-Mar-19           NaN  €500K  €1K   

      Release Clause  Attacking  Crossing  Finishing  Heading Accuracy  \
7979           €1.9M        270        63         59                24   
12982          €975K        278        55         61                57   
15600          €449K        261        51         55                53   

       Short Passing  Volleys  Skill  Dribbling  Curve  FK Accuracy  \
7979              67       57    326         69     59           69   
12982             58       47    258         64     45           39   
15600             62       40    283         57     50           57   

       Long Passing  Ball Control  Movement  Acceleration  Sprint Speed  \
7979             62            67       372            78            79   
12982            52            58       402            84            84   
15600            59            60       336            71            67   

       Agility  Reactions  Balance  Power  Shot Power  Jumping  Stamina  \
7979        79         50       86    270          65       47       60   
12982       91         57       86    328          57       82       71   
15600       70         56       72    304          56       57       75   

       Strength  Long Shots  Mentality  Aggression  Interceptions  \
7979         39          59        247          35             21   
12982        64          54        249          48             33   
15600        61          55        275          60             49   

       Positioning  Vision  Penalties  Composure  Defending  Marking  \
7979            64      62         65         55         79       29   
12982           59      55         54         58         79       26   
15600           57      60         49         60        164       52   

       Standing Tackle  Sliding Tackle  Goalkeeping  GK Diving  GK Handling  \
7979                25              25           54         15           14   
12982               29              24           52          6           14   
15600               57              55           56          5           13   

       GK Kicking  GK Positioning  GK Reflexes  Total Stats  Base Stats  W/F  \
7979            6               7           12         1618         342  3 ★   
12982           9              14            9         1646         356  3 ★   
15600         

In [57]:
fifa['Weight'].unique()

array(['72kg', '83kg', '87kg', '70kg', '68kg', '80kg', '71kg', '91kg',
       '73kg', '85kg', '92kg', '69kg', '84kg', '96kg', '81kg', '82kg',
       '75kg', '86kg', '89kg', '74kg', '76kg', '64kg', '78kg', '90kg',
       '66kg', '60kg', '94kg', '79kg', '67kg', '65kg', '59kg', '61kg',
       '93kg', '88kg', '97kg', '77kg', '62kg', '63kg', '95kg', '100kg',
       '58kg', '183lbs', '179lbs', '172lbs', '196lbs', '176lbs', '185lbs',
       '170lbs', '203lbs', '168lbs', '161lbs', '146lbs', '130lbs',
       '190lbs', '174lbs', '148lbs', '165lbs', '159lbs', '192lbs',
       '181lbs', '139lbs', '154lbs', '157lbs', '163lbs', '98kg', '103kg',
       '99kg', '102kg', '56kg', '101kg', '57kg', '55kg', '104kg', '107kg',
       '110kg', '53kg', '50kg', '54kg', '52kg'], dtype=object)

**Converting weights into a uniform unit of 'kgs' and renaming the 'Weight' column as 'Weight(kgs)'**

In [58]:
def convert_weight(weight):
	if 'kg' in weight:
		return int(weight.strip('kg'))
	else:
		pounds = int(weight.strip('lbs'))
		return round(pounds/2.205)

In [59]:
fifa['Weight'] = fifa['Weight'].apply(convert_weight)

In [60]:
fifa = fifa.rename(columns={'Weight':'Weight(kgs)'})

In [61]:
fifa.sample(2)

ID             Name             LongName  \
11723  256967  M. Kreuzriegler  Martin Kreuzriegler   
4950   180339   N. Schouterden     Nils Schouterden   

                                               photoUrl  \
11723  https://cdn.sofifa.com/players/256/967/21_60.png   
4950   https://cdn.sofifa.com/players/180/339/21_60.png   

                                               playerUrl Nationality  Age  \
11723  http://sofifa.com/player/256967/martin-kreuzri...     Austria   26   
4950   http://sofifa.com/player/180339/nils-schouterd...     Belgium   31   

       ↓OVA  POT                Club     Contract Contract Status  \
11723    64   68  Sandefjord Fotball  2020 ~ 2021        Contract   
4950     70   70           KAS Eupen  2017 ~ 2021        Contract   

      Contract start contract end  contract length(years)    Positions  \
11723           2020         2021                     1.0           CB   
4950            2017         2021                     4.0  LWB, LM, LB   

       Height(cms)  Weight(kgs) Preferred Foot  BOV Best Position     Joined  \
11723          185           77           Left   66            CB  25-Jun-20   
4950           176           65           Left   70           LWB  03-Jul-17   

      Loan Date End  Value Wage Release Clause  Attacking  Crossing  \
11723           NaN  €725K  €1K          €696K        200        26   
4950            NaN  €1.2M  €7K            €2M        286        71   

       Finishing  Heading Accuracy  Short Passing  Volleys  Skill  Dribbling  \
11723         25                65             57       27    198         44   
4950          45                51             65       54    337         67   

       Curve  FK Accuracy  Long Passing  Ball Control  Movement  Acceleration  \
11723     38           27            43            46       284            57   
4950      69           67            67            67       368            75   

       Sprint Speed  Agility  Reactions  Balance  Power  Shot Power  Jumping  \
11723            65       42         63       57    250          36       68   
4950             72       77         66       78    345          78       74   

       Stamina  Strength  Long Shots  Mentality  Aggression  Interceptions  \
11723       56        68          22        224          61             65   
4950        77        48          68        312          64             66   

       Positioning  Vision  Penalties  Composure  Defending  Marking  \
11723           21      37         40         57        192       62   
4950            64      65         53         65        190       67   

       Standing Tackle  Sliding Tackle  Goalkeeping  GK Diving  GK Handling  \
11723               66              64           57         13           11   
4950                64              59           49          8           10   

       GK Kicking  GK Positioning  GK Reflexes  Total Stats  Base Stats  W/F  \
11723          11              13            9         1405         304  3 ★   
4950            9              13            9         1887         391  3 ★   

       SM     A/W     D/W   IR  PAC  SHO  PAS  DRI  DEF  PHY Hits  
11723  2★  Medium  Medium  1 ★   61   27   42   46   64   64    4  
4950   3★    High  Medium  1 ★   73   58   67   69   64   60    2

In [62]:
fifa['W/F'].unique()

array(['4 ★', '3 ★', '5 ★', '2 ★', '1 ★'], dtype=object)

Formatting the weak foot rating of the player and stripping off the (*)

In [63]:
fifa['W/F'] = fifa['W/F'].str.replace(' ★','')

**Changing the datatype of the weakfoot (W/F) column to integer**

In [64]:
fifa['W/F'] = fifa['W/F'].astype(int)

In [66]:
fifa['W/F'].unique()

array([4, 3, 5, 2, 1])

**Ascertain the reason for missing values in the 'Loan Date End' column**

In [67]:
on_loan = fifa[fifa['Contract Status'] == 'On Loan']
on_loan[['Contract','Contract Status','Loan Date End']]

Contract Contract Status Loan Date End
205    Jun 30, 2021 On Loan         On Loan     30-Jun-21
248    Jun 30, 2021 On Loan         On Loan     30-Jun-21
254    Jun 30, 2021 On Loan         On Loan     30-Jun-21
302    Jun 30, 2021 On Loan         On Loan     30-Jun-21
306    Jun 30, 2021 On Loan         On Loan     30-Jun-21
...                     ...             ...           ...
18472  Aug 31, 2021 On Loan         On Loan     31-Aug-21
18571  Jun 30, 2021 On Loan         On Loan     30-Jun-21
18600  Dec 31, 2020 On Loan         On Loan     31-Dec-20
18622  Dec 31, 2020 On Loan         On Loan     31-Dec-20
18680  Dec 31, 2020 On Loan         On Loan     31-Dec-20

[1013 rows x 3 columns]

**As only 1013 players are on loan, rest of them have missing values in the 'Loan Date End' column**

In [72]:
fifa[['Name', 'Hits']]

Name Hits
0               L. Messi  771
1      Cristiano Ronaldo  562
2               J. Oblak  150
3           K. De Bruyne  207
4              Neymar Jr  595
...                  ...  ...
18974             Xia Ao  NaN
18975           B. Hough  NaN
18976        R. McKinley  NaN
18977       Wang Zhen'ao  NaN
18978          Zhou Xiao  NaN

[18979 rows x 2 columns]

There are players in the dataframe who haven't been searched or their search records aren't mentioned in the dataframe. That's why there are missing values in the 'Hits' column which keeps record of number of searches (hits) for a player